contracts/libs/0.8.x/BytecodeStorageV1.sol

```diff
@@ -180,8 +184,27 @@ library BytecodeStorage {
             // store length of data in first 32 bytes
             mstore(data, size)
             // copy code to memory, excluding the deployer-address
-            extcodecopy(_address, add(data, 0x20), dataOffset, size)
+            extcodecopy(_address, add(data, 0x20), _offset, size)
+        }
+    }
+
+    /**
+     * @notice Read a string from contract bytecode
+     * @param _address address of deployed contract with bytecode containing concat(invalid opcode, version, deployer-address, data)
+     * @return data string read from contract bytecode
+     */
+    function readFromBytecode(
+        address _address
+    ) internal view returns (string memory data) {
+        // the dataOffset for the bytecode
+        uint256 dataOffset = _bytecodeDataOffsetAt(_address);
+        // validate that version is readable
+        // note: never expected to throw for reads, as we
+        //       fall back to SSTORE-2 reads optimistically
```

Comment:
Something I would love feedback in during the PR review phase for this is whether or not we should throw an `"Unsupported Version"` error for attempted reads on contracts that we don't think are SSTORE2-written (don't use 0x00 as the offset byte); this feels like more headache than it is worth, and could potentially be annoying for anyone who may have hypothetically minor-ly forked SSTORE2 to have a different "dead byte", but I can see how it would be more "semantically correct" to throw in this case, and force folks to use the manual offset reads for such scenarios... would love thoughts on this and happy to update accordingly.

Reply:
Great call out – updating to use this pattern.